File: app/src/test/java/com/tutv/android/tv_poster_list/TvPosterListPresenterTest.java
package com.tutv.android.tv_poster_list;

import com.tutv.android.domain.Genre;
import com.tutv.android.domain.Series;
import com.tutv.android.repository.SeriesRepository;
import com.tutv.android.ui.tv_poster_list.TvPosterListPresenter;
import com.tutv.android.ui.tv_poster_list.TvPosterListView;
import com.tutv.android.ui.tv_poster_list.tv_poster.TvPosterPresenter;
import com.tutv.android.ui.tv_poster_list.tv_poster.TvPosterView;
import com.tutv.android.utils.schedulers.BaseSchedulerProvider;
import com.tutv.android.utils.schedulers.ImmediateSchedulerProvider;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Single;

import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class TvPosterListPresenterTest {
    TvPosterListView view;
    private SeriesRepository seriesRepository;

    private final int genreId = 1;
    private final String genreName = "Genre";
    private final int pageSize = 10;
    private Genre genreFirstPage;
    private Genre genreSecondPage;
    private Genre genreEmptyPage;

    private TvPosterListPresenter presenter;

    @Before
    public void setup() {
        view = mock(TvPosterListView.class);
        seriesRepository = mock(SeriesRepository.class);
        BaseSchedulerProvider schedulerProvider = new ImmediateSchedulerProvider();

        genreFirstPage = new Genre(genreId, genreName);
        genreSecondPage = new Genre(genreId, genreName);
        genreEmptyPage = new Genre(genreId, genreName);
        List<Series> seriesFirstPage = new ArrayList<>();
        List<Series> seriesSecondPage = new ArrayList<>();
        List<Series> seriesEmptyPage = new ArrayList<>();
        for (int i = 0; i < pageSize; i++) {
            seriesFirstPage.add(new Series());
            seriesSecondPage.add(new Series());
        }
        genreFirstPage.setSeries(seriesFirstPage);
        genreSecondPage.setSeries(seriesSecondPage);
        genreEmptyPage.setSeries(seriesEmptyPage);

        Genre genrePage2 = new Genre(genreId, genreName);
        List<Series> seriesPage2 = new ArrayList<>();
        for (int i = 0; i < pageSize; i++)
            seriesPage2.add(new Series());
        genrePage2.setSeries(seriesPage2);



        presenter = new TvPosterListPresenter(view, seriesRepository, schedulerProvider,
                genreId, genreName, pageSize);
    }

    @Test
    public void givenTheViewWasAttachedThenSetListName() {
        presenter.onViewAttached();

        verify(view).setListName(genreName);
    }

    @Test
    public void givenTheViewWasAttachedThenSeriesAreFetched() {
        when(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage));

        presenter.onViewAttached();

        verify(seriesRepository, timeout(1000).times(1)).getGenreById(genreId, 1, pageSize);
        verify(view, timeout(1000).times(1)).setLoadingStatus(false);
    }

    @Test
    public void givenNextPageIsRequestedThenLoadNextPage() {
        when(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage));
        when(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.just(genreSecondPage));

        presenter.onViewAttached();
        presenter.getNextPage();

        verify(seriesRepository, timeout(1000).times(1)).getGenreById(genreId, 1, pageSize);
        verify(seriesRepository, timeout(1000).times(1)).getGenreById(genreId, 2, pageSize);
        verify(view, timeout(1000).times(1)).setLoadingStatus(true);
        verify(view, timeout(1000).times(2)).setLoadingStatus(false);
    }

    @Test
    public void givenLastPageReachedThenNotifyEndReached() {
        when(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage));
        when(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.just(genreEmptyPage));

        presenter.onViewAttached();
        presenter.getNextPage();

        verify(view, timeout(1000).times(1)).notifyEndReached();
    }

    @Test
    public void givenViewWasAttachedWhenNetworkErrorThenShowError() {
        when(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.error(new Throwable()));

        presenter.onViewAttached();

        verify(view, timeout(1000).times(1)).showLoadError();
    }

    @Test
    public void givenNextPageIsRequestedWhenNetworkErrorThenShowError() {
        when(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage));
        when(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.error(new Throwable()));

        presenter.onViewAttached();
        presenter.getNextPage();

        verify(view, timeout(1000).times(1)).setLoadingStatus(false);
        verify(view, timeout(1000).times(1)).showLoadError();
    }

    @Test
    public void givenViewIsAttachedThenLoadTheCorrectSeries() {
        when(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage));

        TvPosterView tvPosterView = mock(TvPosterView.class);
        TvPosterPresenter tvPosterPresenter = mock(TvPosterPresenter.class);

        when(tvPosterView.getPresenter()).thenReturn(tvPosterPresenter);

        presenter.onViewAttached();

        InOrder inOrder = inOrder(tvPosterPresenter);

        for (int i = 0; i < genreFirstPage.getSeries().size(); i++) {
            presenter.onBindRepositoryRowViewAtPosition(i, tvPosterView);
            inOrder.verify(tvPosterPresenter).bind(genreFirstPage.getSeries().get(i));
        }
        verifyNoMoreInteractions(tvPosterPresenter);
    }

    @Test
    public void givenNextPageIsRequestedThenPagesAreAppendedInCorrectOrder() {
        when(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage));
        when(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.just(genreSecondPage));

        TvPosterView tvPosterView = mock(TvPosterView.class);
        TvPosterPresenter tvPosterPresenter = mock(TvPosterPresenter.class);

        when(tvPosterView.getPresenter()).thenReturn(tvPosterPresenter);

        presenter.onViewAttached();
        presenter.getNextPage();

        int sizeFirstPage = genreFirstPage.getSeries().size();
        int sizeSecondPage = genreSecondPage.getSeries().size();

        InOrder inOrder = inOrder(tvPosterPresenter);

        for (int i = 0; i < sizeFirstPage; i++) {
            presenter.onBindRepositoryRowViewAtPosition(i, tvPosterView);
            inOrder.verify(tvPosterPresenter).bind(genreFirstPage.getSeries().get(i));
        }
        for (int j = 0; j < sizeSecondPage; j++) {
            presenter.onBindRepositoryRowViewAtPosition(sizeFirstPage + j, tvPosterView);
            inOrder.verify(tvPosterPresenter).bind(genreSecondPage.getSeries().get(j));
        }
        verifyNoMoreInteractions(tvPosterPresenter);
    }
}

File: app/src/test/java/com/tutv/android/tv_poster_list/TvPosterListPresenterTest.kt
package com.tutv.android.tv_poster_list

import com.tutv.android.domain.Genre
import com.tutv.android.domain.Series
import com.tutv.android.repository.SeriesRepository
import com.tutv.android.ui.tv_poster_list.TvPosterListPresenter
import com.tutv.android.ui.tv_poster_list.TvPosterListView
import com.tutv.android.ui.tv_poster_list.tv_poster.TvPosterPresenter
import com.tutv.android.ui.tv_poster_list.tv_poster.TvPosterView
import com.tutv.android.utils.schedulers.BaseSchedulerProvider
import com.tutv.android.utils.schedulers.ImmediateSchedulerProvider
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import java.util.*

class TvPosterListPresenterTest {
    private var view: TvPosterListView? = null
    private var seriesRepository: SeriesRepository? = null
    private val genreId = 1
    private val genreName: String? = "Genre"
    private val pageSize = 10
    private var genreFirstPage: Genre? = null
    private var genreSecondPage: Genre? = null
    private var genreEmptyPage: Genre? = null
    private var presenter: TvPosterListPresenter? = null

    @Before
    fun setup() {
        view = Mockito.mock(TvPosterListView::class.java)
        seriesRepository = Mockito.mock(SeriesRepository::class.java)
        val schedulerProvider: BaseSchedulerProvider = ImmediateSchedulerProvider()
        genreFirstPage = Genre(genreId, genreName)
        genreSecondPage = Genre(genreId, genreName)
        genreEmptyPage = Genre(genreId, genreName)
        val seriesFirstPage: MutableList<Series?> = ArrayList()
        val seriesSecondPage: MutableList<Series?> = ArrayList()
        val seriesEmptyPage: MutableList<Series?> = ArrayList()
        for (i in 0 until pageSize) {
            seriesFirstPage.add(Series())
            seriesSecondPage.add(Series())
        }
        genreFirstPage.setSeries(seriesFirstPage)
        genreSecondPage.setSeries(seriesSecondPage)
        genreEmptyPage.setSeries(seriesEmptyPage)
        presenter = TvPosterListPresenter(view, seriesRepository, schedulerProvider,
                genreId, genreName, pageSize)
    }

    @Test
    fun givenTheViewWasAttachedThenSetListName() {
        presenter.onViewAttached()
        Mockito.verify(view).setListName(genreName)
    }

    @Test
    fun givenTheViewWasAttachedThenSeriesAreFetched() {
        Mockito.`when`(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage))
        presenter.onViewAttached()
        Mockito.verify(seriesRepository, Mockito.timeout(1000).times(1)).getGenreById(genreId, 1, pageSize)
        Mockito.verify(view, Mockito.timeout(1000).times(1)).setLoadingStatus(false)
    }

    @Test
    fun givenNextPageIsRequestedThenLoadNextPage() {
        Mockito.`when`(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage))
        Mockito.`when`(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.just(genreSecondPage))
        presenter.onViewAttached()
        presenter.getNextPage()
        Mockito.verify(seriesRepository, Mockito.timeout(1000).times(1)).getGenreById(genreId, 1, pageSize)
        Mockito.verify(seriesRepository, Mockito.timeout(1000).times(1)).getGenreById(genreId, 2, pageSize)
        Mockito.verify(view, Mockito.timeout(1000).times(1)).setLoadingStatus(true)
        Mockito.verify(view, Mockito.timeout(1000).times(2)).setLoadingStatus(false)
    }

    @Test
    fun givenLastPageReachedThenNotifyEndReached() {
        Mockito.`when`(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage))
        Mockito.`when`(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.just(genreEmptyPage))
        presenter.onViewAttached()
        presenter.getNextPage()
        Mockito.verify(view, Mockito.timeout(1000).times(1)).notifyEndReached()
    }

    @Test
    fun givenTheViewWasAttachedWhenNetworkErrorThenShowError() {
        Mockito.`when`(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.error(Throwable()))
        presenter.onViewAttached()
        Mockito.verify(view, Mockito.timeout(1000).times(1)).showLoadError()
    }

    @Test
    fun givenNextPageIsRequestedWhenNetworkErrorThenShowError() {
        Mockito.`when`(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage))
        Mockito.`when`(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.error(Throwable()))
        presenter.onViewAttached()
        presenter.getNextPage()
        Mockito.verify(view, Mockito.timeout(1000).times(1)).setLoadingStatus(false)
        Mockito.verify(view, Mockito.timeout(1000).times(1)).showLoadError()
    }

    @Test
    fun givenTheViewWasAttachedThenLoadTheCorrectSeries() {
        Mockito.`when`(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage))
        val tvPosterView = Mockito.mock(TvPosterView::class.java)
        val tvPosterPresenter = Mockito.mock(TvPosterPresenter::class.java)
        Mockito.`when`(tvPosterView.getPresenter()).thenReturn(tvPosterPresenter)
        presenter.onViewAttached()
        val inOrder = Mockito.inOrder(tvPosterPresenter)
        for (i in genreFirstPage.getSeries().indices) {
            presenter.onBindRepositoryRowViewAtPosition(i, tvPosterView)
            inOrder.verify(tvPosterPresenter).bind(genreFirstPage.getSeries()[i])
        }
        Mockito.verifyNoMoreInteractions(tvPosterPresenter)
    }

    @Test
    fun givenNextPageIsRequestedThenPagesAreAppendedInCorrectOrder() {
        Mockito.`when`(seriesRepository.getGenreById(genreId, 1, pageSize)).thenReturn(Single.just(genreFirstPage))
        Mockito.`when`(seriesRepository.getGenreById(genreId, 2, pageSize)).thenReturn(Single.just(genreSecondPage))
        val tvPosterView = Mockito.mock(TvPosterView::class.java)
        val tvPosterPresenter = Mockito.mock(TvPosterPresenter::class.java)
        Mockito.`when`(tvPosterView.getPresenter()).thenReturn(tvPosterPresenter)
        presenter.onViewAttached()
        presenter.getNextPage()
        val sizeFirstPage = genreFirstPage.getSeries().size
        val sizeSecondPage = genreSecondPage.getSeries().size
        val inOrder = Mockito.inOrder(tvPosterPresenter)
        for (i in 0 until sizeFirstPage) {
            presenter.onBindRepositoryRowViewAtPosition(i, tvPosterView)
            inOrder.verify(tvPosterPresenter).bind(genreFirstPage.getSeries()[i])
        }
        for (j in 0 until sizeSecondPage) {
            presenter.onBindRepositoryRowViewAtPosition(sizeFirstPage + j, tvPosterView)
            inOrder.verify(tvPosterPresenter).bind(genreSecondPage.getSeries()[j])
        }
        Mockito.verifyNoMoreInteractions(tvPosterPresenter)
    }
}